Source/Dafny/DafnyAst.cs

```diff
@@ -114,7 +114,7 @@ public class BuiltIns
     public readonly Dictionary<int, ArrowTypeDecl> ArrowTypeDecls = new Dictionary<int, ArrowTypeDecl>();
     public readonly Dictionary<int, SubsetTypeDecl> PartialArrowTypeDecls = new Dictionary<int, SubsetTypeDecl>();  // same keys as arrowTypeDecl
     public readonly Dictionary<int, SubsetTypeDecl> TotalArrowTypeDecls = new Dictionary<int, SubsetTypeDecl>();  // same keys as arrowTypeDecl
-    readonly Dictionary<int, TupleTypeDecl> tupleTypeDecls = new Dictionary<int, TupleTypeDecl>();
+    readonly Dictionary<object, TupleTypeDecl> tupleTypeDecls = new Dictionary<object, TupleTypeDecl>();
```

Comment:
Can `object` be more specific? For example, is it always a list of booleans?

Will `Dictionary` understand how to compare these keys?

If something else is needed as a key, an alternative would be to treat the array of booleans as a list of binary digits. Then prepend a `1` to those digits and treat as a `BigNumber`. For example, `('a', ghost 'b', ghost 'c')` would be represented by the binary number `1011` (that is, `11` in decimal).

Oh, or better yet, use the strings you already use, like `OGG` for the example above.

Reply:
This code comes from the linear-types branch. They use nested tuples to index the dictionary by a sequence of booleans because C#'s `List` class doesn't provide the correct hash and equality methods.

If we wish to use `List<bool>` for the keys, we can pass to the constructor of the dictionary a comparer like the following (from [here](https://stackoverflow.com/a/10020580/2491528)).
```c#
class ListComparer<T> : IEqualityComparer<List<T>>
{
    public bool Equals(List<T> x, List<T> y)
    {
        return x.SequenceEqual(y);
    }

    public int GetHashCode(List<T> obj)
    {
        int hashcode = 0;
        foreach (T t in obj)
        {
            // etc
        }
        return hashcode;
    }
}
```

Using a `BigNumber` for the keys would also work. Constructing a key would be slower, but the hash and equality operations might be faster.

Reply:
I changed the dictionary to use a `List<bool>` with a custom comparator.
```c#
readonly Dictionary<List<bool>, TupleTypeDecl> tupleTypeDecls = new Dictionary<List<bool>, TupleTypeDecl>(new Dafny.ListComparer<bool>());
```